File: scrape_wechat.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

async function scrapeWeChatArticle(url) {
  console.log('🚀 开始爬取微信文章...');
  console.log(`URL: ${url}`);
  
  // 启动浏览器（headless 模式，服务器环境）
  const browser = await chromium.launch({
    headless: true, // 服务器环境必须使用 headless
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--disable-gpu',
      '--disable-web-security',
      '--disable-features=IsolateOrigins,site-per-process',
      '--window-size=1920,1080'
    ]
  });

  try {
    // 创建浏览器上下文，设置合适的 User-Agent
    const context = await browser.newContext({
      viewport: { width: 1920, height: 1080 },
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 MicroMessenger/7.0.20.1781(0x6700143B) NetType/WIFI MiniProgramEnv/Windows WindowsWechat/WMPF WindowsWechat(0x63090a13) AppType/PC',
      // 模拟微信环境
      extraHTTPHeaders: {
        'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
      }
    });

    const page = await context.newPage();

    // 导航到页面，等待网络空闲
    console.log('📖 正在加载页面...');
    await page.goto(url, {
      waitUntil: 'networkidle',
      timeout: 60000
    });

    // 额外等待时间，确保动态内容完全加载
    console.log('⏳ 等待页面完全加载...');
    await page.waitForTimeout(5000);

    // 等待文章主体出现
    await page.waitForSelector('#js_content, .rich_media_content', {
      state: 'visible',
      timeout: 30000
    });

    console.log('✅ 页面加载完成，开始提取内容...');

    // 提取文章标题
    const title = await page.$eval(
      '#activity-name, h1 rich_media_title, .rich_media_title',
      el => el.textContent?.trim() || ''
    ).catch(() => '');

    // 提取作者信息
    let author = '';
    try {
      author = await page.$eval(
        '#js_name, .rich_media_meta_nickname, span.rich_media_meta_text',
        el => el.textContent?.trim() || ''
      ).catch(() => '');
    } catch (e) {
      author = '未知作者';
    }

    // 提取发布日期
    let publishDate = '';
    try {
      publishDate = await page.$eval(
        'em[id*="publish_time"], .rich_media_meta_text',
        el => el.textContent?.trim() || ''
      ).catch(() => '');
    } catch (e) {
      publishDate = new Date().toLocaleDateString('zh-CN');
    }

    // 提取正文内容
    const content = await page.$eval(
      '#js_content, .rich_media_content',
      el => {
        // 获取所有段落
        const paragraphs = el.querySelectorAll('p, section:not(section section)');
        const textParts = [];
        
        paragraphs.forEach(p => {
          const text = p.textContent?.trim();
          if (text && text.length > 0) {
            textParts.push(text);
          }
        });
        
        return textParts.join('\n\n');
      }
    ).catch(() => '');

    // 提取图片
    const images = await page.$$eval(
      '#js_content img, .rich_media_content img',
      imgs => imgs.map(img => ({
        src: img.getAttribute('data-src') || img.getAttribute('src') || '',
        alt: img.getAttribute('alt') || img.getAttribute('title') || ''
      })).filter(img => img.src && img.src.startsWith('http'))
    ).catch(() => []);

    console.log(`📊 提取完成：标题="${title.substring(0, 30)}...", 作者="${author}", 图片数=${images.length}`);

    // 生成 Markdown 格式
    let markdown = `# ${title}\n\n`;
    markdown += `**作者**: ${author}\n\n`;
    markdown += `**发布日期**: ${publishDate}\n\n`;
    markdown += `**原文链接**: ${url}\n\n`;
    markdown += `---\n\n`;

    // 添加图片
    if (images.length > 0) {
      markdown += `## 图片\n\n`;
      images.forEach((img, index) => {
        markdown += `![${img.alt || `图片${index + 1}`}](${img.src})\n\n`;
      });
      markdown += `---\n\n`;
    }

    // 添加正文
    markdown += `## 正文\n\n`;
    markdown += content;
    markdown += `\n\n---\n\n`;
    markdown += `_本文通过 Playwright 自动爬取于 ${new Date().toLocaleString('zh-CN')}_\n`;

    // 生成摘要
    const summary = generateSummary(title, author, content);

    // 保存到文件
    const outputDir = '/home/zhaog/.openclaw/workspace/memory/';
    const safeTitle = title.substring(0, 50).replace(/[\/\\:*?"<>|]/g, '_') || 'wechat_article';
    const fileName = `${safeTitle}_${Date.now()}.md`;
    const filePath = path.join(outputDir, fileName);

    // 确保目录存在
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    fs.writeFileSync(filePath, markdown, 'utf-8');
    console.log(`💾 文章已保存到: ${filePath}`);

    // 关闭浏览器
    await browser.close();

    return {
      success: true,
      title,
      author,
      publishDate,
      filePath,
      imageCount: images.length,
      contentLength: content.length,
      summary
    };

  } catch (error) {
    console.error('❌ 爬取失败:', error.message);
    await browser.close();
    throw error;
  }
}

function generateSummary(title, author, content) {
  // 提取前 300 字作为摘要
  const preview = content.substring(0, 300);
  const lines = preview.split('\n').filter(line => line.trim().length > 0);
  
  let summary = `📝 **文章核心摘要**\n\n`;
  summary += `**标题**: ${title}\n`;
  summary += `**作者**: ${author}\n\n`;
  summary += `**内容预览**:\n`;
  summary += lines.slice(0, 3).join('\n');
  summary += lines.length > 3 ? '...' : '';
  summary += `\n\n**全文长度**: ${content.length} 字符`;
  
  return summary;
}

// 主程序
(async () => {
  const url = 'https://mp.weixin.qq.com/s/aFDBdv4h9Z_aSYMtgThAAw';
  
  try {
    const result = await scrapeWeChatArticle(url);
    
    console.log('\n✅ 爬取成功！\n');
    console.log(result.summary);
    console.log(`\n📁 保存位置: ${result.filePath}`);
    
  } catch (error) {
    console.error('\n❌ 任务失败:', error.message);
    process.exit(1);
  }
})();

File: archive/scripts/scrape_wechat.py
#!/usr/bin/env python3
"""
微信公众号文章爬取脚本
使用 Playwright 进行真实浏览器操作，处理动态加载内容
"""

import asyncio
from playwright.async_api import async_playwright
from datetime import datetime
import json
import os

async def scrape_wechat_article(url, output_dir):
    """爬取微信公众号文章"""
    
    print(f"🌐 开始爬取：{url}")
    
    async with async_playwright() as p:
        # 启动浏览器（非无头模式可以看到浏览器窗口）
        browser = await p.chromium.launch(
            headless=True,
            args=[
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage',
                '--disable-accelerated-2d-canvas',
                '--disable-gpu'
            ]
        )
        
        # 创建浏览器上下文，设置 User-Agent
        context = await browser.new_context(
            user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 MicroMessenger/7.0.20.1781(0x6700143B) WindowsWechat(0x63090a13) 0.0.1.1901',
            viewport={'width': 1920, 'height': 1080},
            device_scale_factor=1,
            java_script_enabled=True,
            bypass_csp=True
        )
        
        page = await context.new_page()
        
        # 设置额外的请求头
        await page.set_extra_http_headers({
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
            'Connection': 'keep-alive',
        })
        
        try:
            # 导航到页面，等待网络空闲
            print("⏳ 正在加载页面...")
            response = await page.goto(
                url,
                wait_until='networkidle',
                timeout=60000
            )
            
            if response.status != 200:
                print(f"❌ 页面返回状态码：{response.status}")
                return None
            
            # 等待页面完全加载
            print("⏳ 等待页面元素加载...")
            await page.wait_for_timeout(5000)  # 额外等待 5 秒确保动态内容加载
            
            # 等待文章容器出现
            try:
                await page.wait_for_selector('#js_content', timeout=10000)
            except:
                print("⚠️ 未找到标准文章容器，尝试其他选择器...")
            
            # 提取标题
            print("📝 提取标题...")
            title = await page.evaluate('''() => {
                const titleEl = document.querySelector('#activity-name') || 
                               document.querySelector('h1') || 
                               document.querySelector('.rich_media_title');
                return titleEl ? titleEl.textContent.trim() : '未知标题';
            }''')
            
            # 提取作者信息
            print("✍️ 提取作者信息...")
            author = await page.evaluate('''() => {
                const authorEl = document.querySelector('#js_name') || 
                                document.querySelector('.rich_media_meta_nickname') ||
                                document.querySelector('[data-type="author"]');
                return authorEl ? authorEl.textContent.trim() : '未知作者';
            }''')
            
            # 提取发布时间
            print("📅 提取发布时间...")
            publish_date = await page.evaluate('''() => {
                const dateEl = document.querySelector('#publish_time') || 
                              document.querySelector('.rich_media_meta_text');
                return dateEl ? dateEl.textContent.trim() : '';
            }''')
            
            # 提取正文内容
            print("📄 提取正文内容...")
            content_html = await page.evaluate('''() => {
                const contentEl = document.querySelector('#js_content');
                return contentEl ? contentEl.innerHTML : '';
            }''')
            
            # 提取纯文本内容
            content_text = await page.evaluate('''() => {
                const contentEl = document.querySelector('#js_content');
                return contentEl ? contentEl.textContent.trim() : '';
            }''')
            
            # 提取图片信息
            print("🖼️ 提取图片信息...")
            images = await page.evaluate('''() => {
                const imgs = document.querySelectorAll('#js_content img');
                return Array.from(imgs).map(img => ({
                    src: img.getAttribute('data-src') || img.getAttribute('src'),
                    alt: img.getAttribute('alt') || '',
                    width: img.getAttribute('width') || '',
                    height: img.getAttribute('height') || ''
                })).filter(img => img.src);
            }''')
            
            # 关闭浏览器
            await browser.close()
            
            # 生成 Markdown 内容
            print("📝 生成 Markdown 格式...")
            markdown_content = generate_markdown(title, author, publish_date, content_text, images, url)
            
            # 保存文件
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            safe_title = "".join([c for c in title if c.isalnum() or c in ' _-']).strip()[:50]
            filename = f"{timestamp}_{safe_title}.md"
            filepath = os.path.join(output_dir, filename)
            
            os.makedirs(output_dir, exist_ok=True)
            
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(markdown_content)
            
            print(f"✅ 文章已保存至：{filepath}")
            
            # 返回摘要信息
            return {
                'title': title,
                'author': author,
                'publish_date': publish_date,
                'content_length': len(content_text),
                'image_count': len(images),
                'filepath': filepath,
                'summary': content_text[:500] + '...' if len(content_text) > 500 else content_text
            }
            
        except Exception as e:
            print(f"❌ 爬取失败：{str(e)}")
            await browser.close()
            raise

def generate_markdown(title, author, publish_date, content, images, url):
    """生成结构化 Markdown"""
    
    md = f"""# {title}

> **作者**: {author}
> **发布时间**: {publish_date}
> **原文链接**: {url}
> **抓取时间**: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}

---

## 正文内容

{content}

---

## 文章图片 ({len(images)} 张)

"""
    
    for i, img in enumerate(images, 1):
        md += f"![图片 {i}]({img['src']})\n"
    
    md += f"""
---

*此文档由米粒儿自动抓取生成*
"""
    
    return md

async def main():
    url = "https://mp.weixin.qq.com/s/h-mivAS0azlkkkVi4HmzYA"
    output_dir = "/home/zhaog/.openclaw/workspace/memory"
    
    result = await scrape_wechat_article(url, output_dir)
    
    if result:
        print("\n" + "="*60)
        print("📊 文章摘要")
        print("="*60)
        print(f"标题：{result['title']}")
        print(f"作者：{result['author']}")
        print(f"发布时间：{result['publish_date']}")
        print(f"正文字数：{result['content_length']}")
        print(f"图片数量：{result['image_count']}")
        print(f"保存路径：{result['filepath']}")
        print("\n核心内容摘要:")
        print(result['summary'])
        print("="*60)

if __name__ == "__main__":
    asyncio.run(main())
